src/http.cc

```diff
@@ -1384,8 +1384,18 @@ HttpStateData::decodeAndWriteReplyBody()
     MemBuf decodedData;
     decodedData.init();
     httpChunkDecoder->setPayloadBuffer(&decodedData);
-    const bool doneParsing = httpChunkDecoder->parse(inBuf);
+    httpChunkDecoder->syncInputBuffer(inBuf);
+    // Allow httpChunkDecoder be the only owner of input buffer content
+    // to avoid reallocation on cow or and chop operations
+    inBuf = SBuf();
+
+    const bool doneParsing = httpChunkDecoder->parseInput();
     inBuf = httpChunkDecoder->remaining(); // sync buffers after parse
+
+    // Allow HttpStateData be the only owner of input buffer MemBlob
+    // to avoid reallocation on cow or consume/chop operations:
+    httpChunkDecoder->syncInputBuffer(SBuf());
```

Comment:
Unfortunately this is required. This is because the httpChunkDecoder::buf_ and HttpStateData::inBuf points to the same blob. Any append or reserveSpace operation in inBuf will cause reallocation.

What about one of the followings:
  1) Make the TeChunkedParser an standalone class and creates and uses an  Http1::Parser to parse trailers. In the worst case holds an pointer to the HttpStateData::inBuf to parse it, or even better a StringView of inBuf :-) .
   2) Do not create the TeChunkedParser once before starts reading body data but a new temporary TeChunkedParser in HttpStateData::decodeAndWriteReplyBody. The only we need is to store the parser state.
  3) Modify all Http::one::Parser classes to not store the input buffer anywhere.

Reply:
> This is because the httpChunkDecoder::buf_ and HttpStateData::inBuf points to the same blob. Any append [...] operation in inBuf will cause reallocation.

Could you detail this part of your claim, please? For example, I do not understand why the following pseudocode would result in inBuf reallocation:

```C++
    inBuf.reserveSpace(100); // expecting one (re)allocation here, giving us 100+ space bytes
    inBuf.append(..., 10); // added 10 bytes; no reallocation; 90+ space bytes remain
    buf_ = inBuf;
    inBuf.append(..., 20); // added 20 bytes; no reallocation; 70+ space bytes remain
```

Why do you expect the presence of `buf_` to cause reallocation in the _second_ append?

Reply:
I should say "any append or reserveSpace can cause reallocation" . The example you are givving will not cause reallocation. But the following will do:

   ```
    inBuf.reserveSpace(100); // expecting one (re)allocation here, giving us 100+ space bytes
    inBuf.append(..., 50);
    buf_ = inBuf;
    inBuf.consume(..., 50); // inBuf is empty now The store has spaceSize 50
    inBuf.append(..., 80);
```
If the buf_ did not point to the same blob with inBuf we would not need to reallocate in the above example.

Moreover we are doing something like the following:
```
    inBuf.reserveSpace(100); // expecting one (re)allocation here, giving us 100+ space bytes
    inBuf.append(..., 50);
    buf_ = inBuf;
    buf_.consume(..., 50); 
    // buf_ is empty now but continue pointing to the same store which
    // has  spaceSize of 50
    // The following call will reallocate because the attached blob is
    // locked by more than one holders, even if the other holders
    // are empty 
    inBuf.reverseSpace(..., 100); 
```

Probably we can found something. For example if we are empty and there are other lockers of the blob stop using it. But in the other hand it is possible that trying to fix this specific case we are going to hurt something else...

Reply:
> If the buf_ did not point to the same blob with inBuf, we would not need to reallocate in the above example.

Yes, but to get to that "different blob" situation, we would also have an extra memory copying and an extra allocation somewhere earlier! If that is the choice, it is better to have one memory allocation (without copying) than one memory allocation (with copying).

Moreover, if there is a choice between memory allocation (without copying) and copying (without memory allocation), memory allocation (without copying) may be cheaper if that allocation is _pooled_ and there are _many_ bytes to copy.

> For example if we are empty and there are other lockers of the blob stop using it.

Yes, those are the kind of optimizations that I was thinking about when I said that I hope that the high-level code adjustments that this change requests points to are unwelcomed. I hope we can find a way to solve the problem on the lower level.

> But in the other hand it is possible that trying to fix this specific case we are going to hurt something else...

Very true. If possible, our optimizations should be limited to cases where we cannot make things worse, only better. For that, we may have to keep in mind which operations are "better" or "worse" from performance point of view, similar to what I was discussing earlier in this comment.

This is the kind of difficult work that I urged those adding SBuf to Squid to do _before_ adding SBuf. Now we need to to their dirty work. Hopefully, it is not too late.

Reply:
To summarize, I think these specific high-level code changes should be undone (at least for now). We should make internal SBuf changes and changes to inBuf allocation code to accomplish the same overall outcome -- no wasteful reallocations in the targeted use cases. We already have change requests tracking those other two areas so this change request can be resolved if the changes here are undone.

Reply:
Said that, we may need to adjust httpChunkDecoder API or implementation. However, the final version of _this_ code should still be very similar if not identical to the current one:

```C++
const auto doneParsing = httpChunkDecoder->parse(inBuf);
inBuf = httpChunkDecoder->remaining(); // sync buffers after parse
```

... because that (or something very similar) is a natural parsing interface that everybody would want to use. If we add a lot of addiitonal steps, developers will forget to do them or will do them in the wrong order. For example, they will forget to reset inBuf right before calling parse().

Reply:
commit 1a424d27 reverts related changes.